ondemand_agentmanager.go

```diff
@@ -72,60 +77,84 @@ func (m *OnDemandAgentManager) Reconfigure(opts OnDemandAgentManagerReconfigureO
 
 // GetClusterAgent returns the Agent which is not bound to any bucket.
 func (m *OnDemandAgentManager) GetClusterAgent() (*Agent, error) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-
+	m.stateLock.Lock()
 	if m.closed {
+		defer m.stateLock.Unlock()
 		return nil, errors.New("agent manager closed")
 	}
+	defer m.stateLock.Unlock()
 
 	return m.clusterAgent, nil
 }
 
 // GetBucketAgent returns the Agent bound to a specific bucket, creating a new Agent as required.
 func (m *OnDemandAgentManager) GetBucketAgent(ctx context.Context, bucketName string) (*Agent, error) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-
+	m.stateLock.Lock()
 	if m.closed {
+		defer m.stateLock.Unlock()
 		return nil, errors.New("agent manager closed")
 	}
+	defer m.stateLock.Unlock()
 
-	if m.bucketAgents == nil {
-		m.bucketAgents = make(map[string]*Agent)
-	} else {
-		bucketAgent, ok := m.bucketAgents[bucketName]
+	fastMap := m.fastMap.Load()
+	if fastMap != nil {
+		agent, ok := (*fastMap)[bucketName]
 		if ok {
-			return bucketAgent, nil
+			return agent, nil
 		}
 	}
 
-	bucketAgent, err := m.makeAgentLocked(ctx, bucketName)
+	agent, err := m.getBucketAgentSlow(ctx, bucketName)
+	if err != nil {
+		return nil, err
+	}
+
+	return agent, nil
+}
+
+func (m *OnDemandAgentManager) getBucketAgentSlow(ctx context.Context, bucketName string) (*Agent, error) {
+	m.slowLock.Lock()
+	defer m.slowLock.Unlock()
+	if agent, ok := m.slowMap[bucketName]; ok {
+		return agent, nil
+	}
+
+	agent, err := m.makeAgent(ctx, bucketName)
 	if err != nil {
 		return nil, err
 	}
+	m.slowMap[bucketName] = agent
 
-	m.bucketAgents[bucketName] = bucketAgent
+	fastMap := make(map[string]*Agent, len(m.slowMap))
+	for name, agent := range m.slowMap {
+		fastMap[name] = agent
+	}
 
-	return bucketAgent, nil
+	m.fastMap.Store(&fastMap)
+
+	return agent, nil
 }
 
 // Close closes the AgentManager and all underlying Agent instances.
 func (m *OnDemandAgentManager) Close() error {
-	m.lock.Lock()
-	defer m.lock.Unlock()
+	m.stateLock.Lock()
+	defer m.stateLock.Unlock()
+
+	m.opts.Logger.Debug("Closing")
+
+	m.closed = true
+	m.fastMap.Store(nil)
```

Comment:
I think that we might want to 'manually' control the stateLock here (or we can continue to use defer by having some sort of 'closeAndReturnClients' function).  This would allow you to 'steal' the slow-map (and replace m.slowMap with an empty list), and then perform the closing of agents outside the lock.  Right now, calling Close() on the bucket manager can cause GetClient calls to become blocked waiting for the lock to be released (ie: blocked on all the agents shutting down).
Admittedly this is a pretty minor issue though, so I wouldn't be against merging this as-is and fixing that additional issue at some later point in time (through a separate issue, or a separate commit).